### PR TITLE
Ember-Server - ServerHelpers.scala - Refactors

### PR DESCRIPTION
**Pull request by diesalbla**
1. Make "upgradeSocket" into a nested method, fewer parameters.
2. Extract serverFromSocket shared local method
3. Extract a runConnectionAux method.
4. replace match with an if-then-else
5. Extract body of "map" as an "onConnection" function.

### CODE REVIEW

File: ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala [hunk 1]
```diff
@@ -173,109 +173,84 @@ private[server] object ServerHelpers extends ServerHelpersPlatform {
...
+                // Since its not enabled, run connection normally.
+                runConnectionAux(socket, ByteVector.empty)
+              }
+          }
 
-        handler.handleErrorWith { t =>
-          Stream.eval(logger.error(t)("Request handler failed with exception")).drain
-        }
+      handler.handleErrorWith { t =>
+        Stream.eval(logger.error(t)("Request handler failed with exception")).drain
```
Comment (valencik) [inline ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala:247]:
```suggestion
        Stream.exec(logger.error(t)("Request handler failed with exception"))
```

File: ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala [hunk 2]
```diff
@@ -173,109 +173,84 @@ private[server] object ServerHelpers extends ServerHelpersPlatform {
...
+              Stream.exec(H2Server.requireConnectionPreface(socket)) ++
+                Stream.resource(serverFromSocket(connect)).drain
+            case (socket, Some(_)) =>
+              // SSL Connection, not h2, will be http/1.1 but thats not how types align
+              // Prior Knowledge is only allowed over clear where application
+              // protocol has not been agreed via handshake
+              runConnectionAux(socket, ByteVector.empty)
+            case (socket, None) => // Cleartext Protocol
+              if (enableHttp2) {
+                // Http2 Prior Knowledge Check, if prelude is first bytes received tread as http2
```
Comment (valencik) [inline ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala:233]:
```suggestion
                // Http2 Prior Knowledge Check, if prelude is first bytes received treat as http2
```

File: ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala [hunk 3]
```diff
@@ -173,109 +173,84 @@ private[server] object ServerHelpers extends ServerHelpersPlatform {
...
-                      onWriteFailure,
-                      createRequestVault,
-                      webSocketKey,
-                      ByteVector.empty,
-                      enableHttp2,
-                    ).drain
+
+    def upgradeSocket(socketInit: Socket[F]): Resource[F, (Socket[F], Option[String])] =
+      tlsInfoOpt match {
+        case None => (socketInit, Option.empty[String]).pure[Resource[F, *]]
```
Comment (valencik) [inline ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala:179]:
```suggestion
        case None => Resource.pure(socketInit, None)
```

Maybe the `.pure` style is preferred? I don't know.
But `Resource.pure(socketInit, None)` feels more straight forward to me.

**Review by valencik — APPROVED** [review]
_Unofficial review_

I like this. I think breaking things up is nice and there's some good de-duplication / simplification here. 👍

File: ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala (hunk 3 above)
Comment (diesalbla) [reply ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala:179]: Me too. Do not much like dealing with asterisks.

File: ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala [hunk 4]
```diff
@@ -173,109 +173,84 @@ private[server] object ServerHelpers extends ServerHelpersPlatform {
...
-                      onWriteFailure,
-                      createRequestVault,
-                      webSocketKey,
-                      ByteVector.empty,
-                      enableHttp2,
-                    ).drain
+
+    def upgradeSocket(socketInit: Socket[F]): Resource[F, (Socket[F], Option[String])] =
+      tlsInfoOpt match {
+        case None => Resource.pure(socketInit, None)
```
Comment (hamnis) [inline ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala:179]: fails on 2.12: https://github.com/http4s/http4s/actions/runs/3826329960/jobs/6510029546#step:10:105

Comment (valencik) [reply ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala:179]: Woops, sorry, should have had more parens:
```suggestion
        case None => Resource.pure((socketInit, None))
```
tested locally, worked with 2.12.17

**Pull request closed without merging**